---
output:
  pdf_document: default
geometry: "left=2cm,right=2cm,bottom=2cm,headheight=48pt"
params: 
    set_subtitle: "Relic_samples"
    snp_coverage_file: "/Users/lamnidis/mount/eager_outputs/2020-05-03-relic_samples/genotyping/double_eigenstrat_coverage.txt"
    sex_det_file: "/Users/lamnidis/mount/eager_outputs/2020-05-03-relic_samples/sex_determination/SexDet.txt"
    stats_table: "/Users/lamnidis/mount/eager_outputs/2020-05-03-relic_samples/multiqc/multiqc_data/multiqc_general_stats.txt"
    poseidon_geno: "/Users/lamnidis/mount/poseidon_packages/2020-05-03-relic_samples/2020-05-03-relic_samples.geno"
    poseidon_snp: "/Users/lamnidis/mount/poseidon_packages/2020-05-03-relic_samples/2020-05-03-relic_samples.snp"
    poseidon_ind: "/Users/lamnidis/mount/poseidon_packages/2020-05-03-relic_samples/2020-05-03-relic_samples.ind"
    poseidon_janno: "/Users/lamnidis/mount/forged_packages/microscope_pca/microscope_pca.janno"
    read_results_fn: "/Users/lamnidis/mount/automated_analysis/2020-05-03-relic_samples/read/2020-05-03-relic_samples.read.txt"
    read_plot_fn: "/Users/lamnidis/mount/automated_analysis/2020-05-03-relic_samples/read/2020-05-03-relic_samples.read.plot.pdf"
    pmmr_results_fn: "/Users/lamnidis/mount/automated_analysis/2020-05-03-relic_samples/pmmr/2020-05-03-relic_samples.pmmr.txt"
    pca_bg_annotation: "/Users/lamnidis/Software/github/Schiffels-Popgen/MICROSCOPE-processing-pipeline/project_reports/assets/bg_annotation.txt"
    we_evec_fn: "/Users/lamnidis/mount/automated_analysis/microscope_pca/West_Eurasian_pca.evec"
    we_eval_fn: "/Users/lamnidis/mount/automated_analysis/microscope_pca/West_Eurasian_pca.eval"
    eu_evec_fn: "/Users/lamnidis/mount/automated_analysis/microscope_pca/Europe_only_pca.evec"
    eu_eval_fn: "/Users/lamnidis/mount/automated_analysis/microscope_pca/Europe_only_pca.eval"
    cred_file: "/Users/lamnidis/Software/github/Schiffels-Popgen/MICROSCOPE-processing-pipeline/.credentials"
    report_date: "18-05-2022"
    logo_file: "/Users/lamnidis/Software/github/Schiffels-Popgen/MICROSCOPE-processing-pipeline/project_reports/assets/MICROSCOPE_logo_blue.pdf"
    mpi_logo_file: "/Users/lamnidis/Software/github/Schiffels-Popgen/MICROSCOPE-processing-pipeline/project_reports/assets/EVA_logo.png"
title: "Preliminary data report"
subtitle: "`r params$set_subtitle`"
header-includes: 
 \usepackage{fancyhdr}
 \fancyhf{}
 \fancypagestyle{plain}{\pagestyle{fancy}}
 \pagestyle{fancy}
 \fancyhead[R]{\includegraphics[height = 1.5cm]{`r params$mpi_logo_file`}}
 \fancyfoot{} 
 \cfoot{Report generated on `r params$report_date`}
 \input{header}
 \renewcommand{\footrulewidth}{0.4pt}
 \usepackage{helvet}
 \usepackage[T1]{fontenc}
 \renewcommand{\familydefault}{\sfdefault}
---
\begin{center}
ERC-Project MICROSCOPE:

Zooming into the Population History of Iron Age Europe with Rare Genetic Variants.

Sample Report.

Project lead: Dr. Stephan Schiffels

Scientific Staff: Dr. Thiseas C. Lamnidis and Dr. Angela MÃ¶tsch

Date of Report: `r params$report_date`

\end{center}

```{r echo=FALSE, fig.align='center'}
knitr::include_graphics(params$logo_file)
```

```{r load_libraries, message=FALSE, warning=FALSE, include=FALSE}
## Load required packages
library(stringr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(readr)
library(magrittr)
# library(knitr)
library(kableExtra)
library(forcats)
library(maps)
library(scales)
library(kableExtra)
library(ggrepel)
if(!require('remotes')) install.packages('remotes')
if(!require('sidora.core')) remotes::install_github("sidora-tools/sidora.core")
if(!require('poseidonR')) remotes::install_github('poseidon-framework/poseidonR')
library(sidora.core)
library(poseidonR)
require(RMariaDB)
# require(admixr)

options(scipen=2)
## Minimum number of SNPs to be kept for analysis
snp_cov_cutoff=10000

## Cutoffs for genetic sex calls and expected ranges in sexdet plot.
#### Females
xx_xmin <- 0.7
xx_xmax <- 1.2
xx_ymin <- 0.0
xx_ymax <- 0.1
#### Males
xy_xmin <- 0.2
xy_xmax <- 0.6
xy_ymin <- 0.3
xy_ymax <- 0.6

sex_cutoffs <- tibble("Expected ranges"=c("XX", "XY"),
                      xmin=c(xx_xmin, xy_xmin),
                      xmax=c(xx_xmax, xy_xmax),
                      ymin=c(xx_ymin, xy_ymin),
                      ymax=c(xx_ymax, xy_ymax))

## Function to infer genetic sex from x/y rates
infer_sex <- function(sex_det, sex_cutoffs){
  ## Convert to a matrix with rownames so code is more legible and to avoid needing to pull() values from the tibbles.
  limits <- sex_cutoffs %>% tibble::column_to_rownames(var="Expected ranges") %>% as.matrix()
  sex_det <- sex_det %>%
    mutate("Genetic Sex" = ifelse(
      `x-rate` > limits["XX","xmin"] &
        `x-rate` < limits["XX","xmax"] &
        `y-rate` > limits["XX","ymin"] &
        `y-rate` < limits["XX","ymax"], "F",
      ifelse(`x-rate` > limits["XY","xmin"] &
               `x-rate` < limits["XY","xmax"] &
               `y-rate` > limits["XY","ymin"] &
               `y-rate` < limits["XY","ymax"], "M",
             "Unknown"))) %>%
    select(`Sample`, `Genetic Sex`)
  return(sex_det)
}
```

\
\
\
\
\
\
\hrule \relax
The following report summarises the preliminary archaeogenetic results for individuals from the following sites:
```{r echo=FALSE}
con <- get_pandora_connection(params$cred_file)
dataset_inds <- read_tsv(params$poseidon_ind, col_names=c("Ind","Sex","Site_Id"), col_types='ccc') %>%
  group_by(Site_Id) %>%
  summarise(.groups='keep', "Number of Individuals"=n())

site_info <- get_df("TAB_Site", con) %>%
  filter(`site.Full_Site_Id` %in% dataset_inds$Site_Id) %>%
  select(`site.Full_Site_Id`, `site.Name`, `site.Longitude`, `site.Latitude`)
site_table <- full_join(site_info, dataset_inds, by=c("site.Full_Site_Id"="Site_Id")) %>%
  rename(
    Site_ID = site.Full_Site_Id,
    Name = site.Name,
    Longitude = site.Longitude,
    Latitude = site.Latitude
  )
site_table %>%
  kable(longtable = T, format.args=list(big.mark = ',', digits=3))  %>%
  kable_styling(full_width=F, latex_options=c("hold_position", "repeat_header"))

```

\newpage
# Descriptive statistics of sequencing data

The table below contains some descriptive statistics on the sequencing data. The meaning of each of these statistics is as follows:

* **Sequenced Reads**: The total number of DNA fragments that were sequenced.
* **Mapped Reads**: The number of human DNA fragments.
* **Endogenous DNA (%)**: The percent of the 'Sequenced Reads' that are identified as human DNA. 
* **% of Sequence Duplication**: The percent of 'Mapped Reads' that are laboratory duplicates of other DNA fragments
* **Damage 1st bp 5'**: The proportion of DNA fragments that show evidence of DNA damage on the 1st base pair (bp) of the 5' end of the DNA molecule.

The proportion of Endogenous DNA is a measure of human DNA preservation. Libraries with less than 0.1% Endogenous DNA are not processed further. Those libraries that pass this threshold are either sequenced deeper or undergo in-solution capture.  

A high percent of sequence duplication indicates that we have sequenced most of the unique DNA fragments in the DNA library, meaning that any further sequencing would provide diminishing returns.

The presence of DNA damage helps confirm that the DNA in a sequenced library is in fact ancient. Given the library preparation protocol used in MPI-EVA, for iron-age material from Europe we generally expect to see roughly 8-10% of DNA damage on the 1st bp of the 5\' end of the DNA molecules. This value can vary considerably based on the specific burial conditions in each site.

Libraries with low proportion of DNA damage (Damage 1st bp 5\') and high endogenous DNA are potentially contaminated with present-day DNA molecules. 

```{r load_general_stats, echo=FALSE, message=FALSE}
library_info <- read_tsv(params$stats_table) %>%
  filter(!grepl("_S0_",Sample)) %>%
  filter(grepl("TF|SG",Sample)) %>%
  select(Sample,
    "Sequenced Reads"=`Samtools Flagstat (pre-samtools filter)_mqc-generalstats-samtools_flagstat_pre_samtools_filter-flagstat_total`,
    "Mapped Reads" = `Samtools Flagstat (pre-samtools filter)_mqc-generalstats-samtools_flagstat_pre_samtools_filter-mapped_passed`,
    "Endogenous DNA (%)"= `endorSpy_mqc-generalstats-endorspy-endogenous_dna`,
    "% of Sequence Duplication" = `Picard_mqc-generalstats-picard-PERCENT_DUPLICATION`,
    "Damage 1st bp 5'" = `DamageProfiler_mqc-generalstats-damageprofiler-5_Prime1`
         ) %>%
  mutate(`% of Sequence Duplication`=`% of Sequence Duplication`*100) %>%
  arrange(Sample)
library_info %>% kable(longtable = TRUE, format.args=list(big.mark = ',', digits=3))  %>% kable_styling(full_width=T, latex_options=c("hold_position", "repeat_header"))
```
\newpage

# Sex determination and SNP coverage

The table below includes the sample name, archaeological ID, genetic sex and SNP coverage of each individual. Genetic Sex is abbreviated as follows:

* **M** : Male
* **F** : Female
* **Unknown** (usually due to low coverage.)

Individuals that have too few covered SNPs (< `r format(snp_cov_cutoff, digits = 0, big.mark = " ", decimal.mark = ",")`) for population genetic analysis are shown in red.

```{r gather snp coverage data, echo=FALSE, message=FALSE}
## Load and parse coverage stats
snp_cov<-read_tsv(params$snp_coverage_file, col_names=c("#Sample", "#SNPs_Covered", "#SNPs_Total"), col_types='cii', skip=1) %>%
  rename("Sample"="#Sample","# SNPs Covered" = "#SNPs_Covered", "Out of a maximum of" = "#SNPs_Total") %>%
  arrange(Sample) ## Sort by individual ID so bad_cov gets the correct indices for highlighting in the table.


bad_cov <- which(snp_cov$`# SNPs Covered` < snp_cov_cutoff)
good_cov_inds <- snp_cov$Sample[which(snp_cov$`# SNPs Covered` >= snp_cov_cutoff)]

```

```{r gather sexdet data, message=FALSE, include=FALSE}
## Set up filter string to only keep sexdet from the strandedness of interest
keep_only <- ifelse(
    str_detect(basename(params$snp_coverage_file), "double"),
    "doublestrand",
    "singlestrand"
    )

## Load and parse sex determination stats.
sex_det <- read_tsv(params$sex_det_file, col_names=c("#Sample", "#SnpsAut", "#SNPsX", "#SnpsY", "NrAut", "NrX", "NrY", "x-rate", "y-rate"), col_types='ciiiiiidd', skip=1) %>%
  filter(grepl(keep_only, `#Sample`)) %>% ## When both ssDNA and dsDNA libraries exist, only keep the strandedness of interest
  separate(`#Sample`, into=c("Sample"), sep="[._]", extra='drop') %>%
  select(`Sample`, `x-rate`, `y-rate`) %>%
  infer_sex(sex_cutoffs=sex_cutoffs)
```

```{r arch_IDs from pandora, message=FALSE, include=FALSE}
## Load ArchIDs from pandora using sidora
pandora_data <- get_df("TAB_Individual", con) %>%
  filter(`individual.Full_Individual_Id` %in% sex_det$Sample) %>%
  select(`individual.Full_Individual_Id`, `individual.Archaeological_ID`) %>%
  rename("Sample" = `individual.Full_Individual_Id`, "Archaeological ID" = `individual.Archaeological_ID`)
RMariaDB::dbDisconnect(con)
```

```{r print coverage_sexdet_arch_id table, echo=FALSE, message=FALSE}
sex_det_pandora <- full_join(sex_det, pandora_data, by="Sample")
snp_cov_sex_pandora <- full_join(snp_cov, sex_det_pandora, by="Sample") %>% select(`Sample`, `Archaeological ID`, `Genetic Sex`, `# SNPs Covered`, `Out of a maximum of`) %>%
  arrange(Sample)

snp_cov_sex_pandora %>% kable(longtable = TRUE, format.args=list(big.mark = ',')) %>% kable_styling(latex_options = c("hold_position", "repeat_header")) %>%
  row_spec(bad_cov, bold = F, color = "red")
```
**These results are liable to change with any addition of further sequencing data.**
\newpage

The following figure shows the relative coverage of each individual on the X and Y chromosomes, and the associated error bars for each measure, as well as the expected position for males and females on such a plot. The expected ranges displayed here were used as the cutoffs for the inference of genetic sex shown in the table on page 2.

```{r sexdet plot, echo=FALSE, fig.height=7, fig.width=6, message=FALSE, warning=FALSE}
sex_det <- read_tsv(params$sex_det_file, col_names=c("#Sample", "#SnpsAut", "#SNPsX", "#SnpsY", "NrAut", "NrX", "NrY", "x-rate", "y-rate", "Err(x-rate)", "Err(y-rate)"), col_types='ciiiiiidddd', skip=1) %>%
  separate(`#Sample`, into=c("Sample"), sep="[._]", extra='drop') %>%
  mutate(xmin=`x-rate`-`Err(x-rate)`,
         xmax=`x-rate`+`Err(x-rate)`,
         ymin=`y-rate`-`Err(y-rate)`,
         ymax=`y-rate`+`Err(y-rate)`,
         `total_snps`=NrAut+NrY+NrX
  ) %>%
  select(Sample, x=`x-rate`, xmin, xmax, y=`y-rate`, ymin, ymax, "Total Snps"=total_snps)

ggplot(sex_det) +
  geom_point(aes(x=x, y=y, alpha=`Total Snps`), size=2.5) +
  geom_rect(data=sex_cutoffs, aes(xmin=xmin, xmax=xmax, ymin=ymin,ymax=ymax, fill=`Expected ranges`), alpha=0.4) +
  geom_errorbar(aes(x=x, ymin=ymin, ymax=ymax, alpha=`Total Snps`), width=0) +
  geom_errorbarh(aes(y=y, xmin=xmin, xmax=xmax, alpha=`Total Snps`), height=0) +
  geom_text_repel(aes(label=Sample, x=x, y=y), max.overlaps = 50, size=2) +
  scale_alpha(range=c(0.3,1), labels=function(x) {format(x, big.mark = ",", decimal.mark = ".", scientific = F)}) +
  theme_classic() +
  theme(legend.position = 'bottom',
        legend.direction="horizontal",
        legend.margin=margin(r=20,l=20),
        legend.box='vertical'
        ) +
  xlab("Relative coverage on the X chromosome") +
  ylab("Relative coverage on the Y chromosome") +
  ggtitle("Genetic sex determination plot") +
  coord_cartesian(xlim=c(0,1.2), ylim=c(0,0.8)) +
  expand_limits(x=c(0,1.2),y=c(0,0.8))
```

\newpage

# Genetic relationships between individuals
```{r import_read_results, echo=FALSE}
pull_arch_id <- function(pandora_data, IID) {
  x <- NULL
  for (iid in IID){
    x <- c(x,dplyr::case_when(
      iid %in% pandora_data$Sample ~ pandora_data %>% filter(`Sample` == iid) %>% pull(`Archaeological ID`),
      TRUE ~ NA_character_
    ))
  }
  return(x)
}

## If not enough samples pass coverage cutoffs, then the read results will not exist. Add a check.
if (file.exists(params$read_results_fn) && file.exists(params$read_plot_fn)) {
  read_results <- read_tsv(params$read_results_fn, col_types = 'ccdd') %>%
    mutate(
      IndA_ID = substr(PairIndividuals, 0,6),
      IndB_ID = substr(PairIndividuals, 7,12),
      IndA = pull_arch_id(pandora_data, IndA_ID),
      IndB = pull_arch_id(pandora_data, IndB_ID)
    ) %>%
    select(IndA_ID, IndB_ID, IndA, IndB, Relationship, Z_upper, Z_lower) %>%
    filter(Relationship != "Unrelated")
    
  show_kinship <- nrow(read_results) > 0
  nothing_to_show <- F
} else {
  show_kinship <- F
  nothing_to_show <- T
}
```
\newif\ifkinship
  \newif\ifnada
\kinship`r ifelse(show_kinship, 'true', 'false')`
\ifkinship

The following table shows all the genetic relationships identified between the individuals presented in this report. The pair of related individuals, their respective archaeological IDs (if any), as well as the inferred genetic relationship is shown. Any pairs of individuals not in the table lack any evidence of genetic kinship to the limit of our resolution. The two final columns capture the confidence of the proposed relationship, by comparing if the proposed relationship is significantly more likely than the relationship that is further by one degree (`Z_upper`), and the relationship that is closer by one degree (`Z_lower`). ZScores below 3 are not statistically significant (p>0.01).

The table is followed by a graphical representation of these results, showing the confidence around each proposed relationship as an error bar. This figure also summarises results for unrelated individuals that are not mentioned in the table.

```{r print_read_table, echo=FALSE, eval=show_kinship}
read_results %>% kable(longtable = TRUE, format.args=list(big.mark = ',')) %>% kable_styling(latex_options = c("hold_position", "repeat_header"))
```
![read_plot](`r params$read_plot_fn`)
\else
  \nada`r ifelse(nothing_to_show, 'true', 'false')`
  \ifnada
Not enough individuals with high enough SNP coverage are available for this analysis. Genetic kinship analysis requires a minimum of 2 individuals with sufficient SNP coverage.
  \else
Genetic analysis revealed no evidence of genetic kinship among the analysed individuals.
  \fi
\fi

\newpage

# Principal component analysis with West Eurasian populations
```{r import PCA data, echo=FALSE, message=FALSE, warning=FALSE}
pop_labels <- read_tsv(params$pca_bg_annotation, col_names = c("pop","label","region","colour"), col_types='ccci')
pca_evec <- readr::read_fwf(params$we_evec_fn,
                          fwf_widths(
                            c(21,11,rep(12,times=9),17),
                              col_names=c("ind","pc1","pc2","pc3","pc4","pc5","pc6","pc7","pc8","pc9","pc10","pop")
                            ),
                          skip=1,
                          trim_ws=T,
                          col_select=c("ind","pc1","pc2","pop"), col_types='cddc'
                          )

bg_plot_data <- inner_join(pca_evec, pop_labels, by="pop")
pca_samples <- pca_evec %>% filter(ind %in% good_cov_inds)

pca_eval <- read_tsv(params$we_eval_fn, col_names = "eigenvalues", col_types='d')

pca_samples <- pca_samples %>% left_join(snp_cov, by=c("ind"="Sample")) %>% select(ind,pc1,pc2,pop, n_snps=`# SNPs Covered`)

## Infer PC orientation and flip if necessary
Chuvash_ind <- bg_plot_data %>% filter(ind == "Ttr-481")
BedouinB_ind <- bg_plot_data %>% filter(ind == "HGDP00636")
if (Chuvash_ind$pc1 > BedouinB_ind$pc1) {
  bg_plot_data <- bg_plot_data %>% mutate(pc1 = -pc1)
  pca_samples <- pca_samples %>% mutate(pc1 = -pc1)
}
if (Chuvash_ind$pc2 < BedouinB_ind$pc2) {
  bg_plot_data <- bg_plot_data %>% mutate(pc2 = -pc2)
  pca_samples <- pca_samples %>% mutate(pc2 = -pc2)
}
show_pca <- nrow(pca_samples) > 0
```
\newif\ifpca
\pca`r ifelse(show_pca, 'true', 'false')`
\ifpca
The following plot shows how ancient individuals  with sufficient genomic coverage (in black) relate to present-day West Eurasian populations. The distance between individuals roughly correlates with genetic distances. The percent of genetic variation explained by each axis is shown in parentheses.

The colour used in the plotting of present-day individuals indicates rough geographic regions, as follows:

* **Caucasus**: \textcolor[HTML]{F8766D}{red}
* **Near East**: \textcolor[HTML]{A3A500}{olive}
* **South Europe**: \textcolor[HTML]{00B0F6}{light blue}
* **Central/North Europe**: \textcolor[HTML]{E76BF3}{magenta}
* **Northeastern/Eastern Europe**: \textcolor[HTML]{00BF7D}{green}

In the next page you will find a zoomed-in plot of the same PCA results, focusing on Europe, as well as a table listing all the populations in each of the five geographic regions listed above.


```{r plot pca, echo=FALSE, fig.height=7, fig.width=7, message=FALSE, warning=FALSE, eval=nrow(pca_samples)>0}
ggplot() +
  coord_equal() +
  geom_text(data=bg_plot_data, aes(x=pc1, y=pc2, label=label, col=factor(colour)), alpha=0.4, size=2, show.legend=F) +
  theme_minimal() +
  geom_point(data=pca_samples, aes(x=pc1, y=pc2, alpha=n_snps)) +
  geom_text_repel(data=pca_samples, aes(x=pc1, y=pc2, label=ind), max.overlaps = Inf, size=2.5) +
  scale_alpha_continuous(
    range=c(0.2,1),
    limits=c(snp_cov_cutoff, 80000),
    labels=function(x) {paste0("at least ", format(x, digits = 0, big.mark = " ", scientific = F)," SNPs")},
    name="SNP Coverage"
    ) +
  scale_color_manual(values=c(
    "1" = "#F8766D",
    "2" = "#A3A500",
    "3" = "#00BF7D",
    "4" = "#00B0F6",
    "5" = "#E76BF3"
    )) +
  # geom_point(data=pca_samples, aes(x=pc1, y=pc2, shape=ind)) +
  xlab(paste0("PC1 (",round((pca_eval$eigenvalues[1]/sum(pca_eval$eigenvalues))*100,digits=3),"%)")) +
  ylab(paste0("PC2 (",round((pca_eval$eigenvalues[2]/sum(pca_eval$eigenvalues))*100,digits=3),"%)")) +
  labs(shape = "Ancient Individual")
  
```
\newpage
```{r plot zoomed pca, echo=FALSE, fig.height=7, fig.width=7, message=FALSE, warning=FALSE, eval=nrow(pca_samples)>0}
ggplot() +
  coord_equal(xlim=c(-0.05,0.03), ylim=c(-0.06,0.02)) +
  geom_text(data=bg_plot_data, aes(x=pc1, y=pc2, label=label, col=factor(colour)), alpha=0.4, size=2, show.legend=F) +
  theme_minimal() +
  geom_point(data=pca_samples, aes(x=pc1, y=pc2, alpha=n_snps)) +
  geom_text_repel(data=pca_samples, aes(x=pc1, y=pc2, label=ind), max.overlaps = Inf, size=2.5) +
  scale_alpha_continuous(
    range=c(0.2,1),
    limits=c(snp_cov_cutoff, 80000),
    labels=function(x) {paste0("at least ", format(x, digits = 0, big.mark = " ", scientific = F)," SNPs")},
    name="SNP Coverage"
    ) +
    scale_color_manual(values=c(
    "1" = "#F8766D",
    "2" = "#A3A500",
    "3" = "#00BF7D",
    "4" = "#00B0F6",
    "5" = "#E76BF3"
    )) +
  # geom_point(data=pca_samples, aes(x=pc1, y=pc2, shape=ind)) +
  xlab(paste0("PC1 (",round((pca_eval$eigenvalues[1]/sum(pca_eval$eigenvalues))*100,digits=3),"%)")) +
  ylab(paste0("PC2 (",round((pca_eval$eigenvalues[2]/sum(pca_eval$eigenvalues))*100,digits=3),"%)")) +
  labs(shape = "Ancient Individual")
  
```

```{r plot pca legend, echo=FALSE, message=FALSE, warning=FALSE, eval=nrow(pca_samples)>0}

## Prepare list of populations for text
pops_for_text <- pop_labels %>%
  group_by(region) %>%
  summarise(pops=paste(pop, collapse=", ")) %>%
  mutate(
    Region=case_when(
      region == "Caucasus" ~ "Caucasus",
      region == "NEast" ~ "Near East",
      region == "NEEur" ~ "Northeastern/Eastern Europe",
      region == "NEur" ~ "Central/North Europe",
      region == "SEur" ~ "South Europe"
    )
  ) %>%
  select(
    Region,
    Populations = pops
  )
  
pops_for_text %>%
  kable(longtable = TRUE, format.args=list(big.mark = ',')) %>%
  kable_styling(latex_options = c("hold_position")) %>%
  column_spec(., 1, width="3.5cm") %>%
  column_spec(., 2, width="13cm") %>%
  row_spec(1, color="#F8766D") %>%
  row_spec(2, color="#A3A500") %>%
  row_spec(3, color="#00BF7D") %>%
  row_spec(4, color="#E76BF3") %>%
  row_spec(5, color="#00B0F6")
```

\newpage
# Principal component analysis with European populations

The following plot shows how ancient individuals  with sufficient genomic coverage (in black) relate to present-day European populations. The distance between individuals roughly correlates with genetic distances. The percent of genetic variation explained by each axis is shown in parentheses.

The colour used in the plotting of present-day individuals indicates rough geographic regions, as follows:

* **South Europe**: \textcolor[HTML]{00B0F6}{light blue}
* **Central/North Europe**: \textcolor[HTML]{E76BF3}{magenta}
* **Northeastern/Eastern Europe**: \textcolor[HTML]{00BF7D}{green}

Below the plot you will also find a table listing all the populations in each of the three geographic regions listed above. The populations in each region are almost identical to the last plot, except the "Chuvash" population has been removed from the "Northeastern/Eastern Europe" region.

```{r import PCA2 data, echo=FALSE, message=FALSE, warning=FALSE, eval=nrow(pca_samples)>0}
eu_pop_labels <- read_tsv(params$pca_bg_annotation, col_names = c("pop","label","region","colour"), col_types='ccci') %>%
  filter(region %in% c("NEEur", "SEur", "NEur"), pop != "Chuvash")
eu_pca_evec <- readr::read_fwf(params$eu_evec_fn,
                          fwf_widths(
                            c(21,11,rep(12,times=9),17),
                              col_names=c("ind","pc1","pc2","pc3","pc4","pc5","pc6","pc7","pc8","pc9","pc10","pop")
                            ),
                          skip=1,
                          trim_ws=T,
                          col_select=c("ind","pc1","pc2","pop"), col_types='cddc'
                          )

eu_bg_plot_data <- inner_join(eu_pca_evec, eu_pop_labels, by="pop")
eu_pca_samples <- eu_pca_evec %>% filter(ind %in% good_cov_inds)

eu_pca_eval <- read_tsv(params$eu_eval_fn, col_names = "eigenvalues", col_types='d')

eu_pca_samples <- eu_pca_samples %>% left_join(snp_cov, by=c("ind"="Sample")) %>% select(ind,pc1,pc2,pop, n_snps=`# SNPs Covered`)

## Infer PC orientation and flip if necessary
Cypriot_ind <- eu_bg_plot_data %>% filter(ind == "CYP19")
Lithuanian_ind <- eu_bg_plot_data %>% filter(ind == "lithuania10")
if (Cypriot_ind$pc1 > Lithuanian_ind$pc1) {
  eu_bg_plot_data <- eu_bg_plot_data %>% mutate(pc1 = -pc1)
  eu_pca_samples <- eu_pca_samples %>% mutate(pc1 = -pc1)
}
if (Cypriot_ind$pc2 < Lithuanian_ind$pc2) {
  eu_bg_plot_data <- eu_bg_plot_data %>% mutate(pc2 = -pc2)
  eu_pca_samples <- eu_pca_samples %>% mutate(pc2 = -pc2)
}
```

```{r plot pca2, echo=FALSE, fig.height=7, fig.width=7, message=FALSE, warning=FALSE, eval=nrow(pca_samples)>0}
ggplot() +
  coord_equal() +
  geom_text(data=eu_bg_plot_data, aes(x=pc2, y=pc1, label=label, col=factor(colour)), alpha=0.4, size=2, show.legend=F) +
  theme_minimal() +
  geom_point(data=eu_pca_samples, aes(x=pc2, y=pc1, alpha=n_snps)) +
  geom_text_repel(data=eu_pca_samples, aes(x=pc2, y=pc1, label=ind), max.overlaps = Inf, size=2.5) +
  scale_alpha_continuous(
    range=c(0.2,1),
    limits=c(snp_cov_cutoff, 80000),
    labels=function(x) {paste0("at least ", format(x, digits = 0, big.mark = " ", scientific = F)," SNPs")},
    name="SNP Coverage"
    ) +
    scale_color_manual(values=c(
    "1" = "#F8766D",
    "2" = "#A3A500",
    "3" = "#00BF7D",
    "4" = "#00B0F6",
    "5" = "#E76BF3"
    )) +
  # geom_point(data=pca_samples, aes((x=pc2, y=pc1, shape=ind)) +
  ylab(paste0("PC1 (",round((eu_pca_eval$eigenvalues[1]/sum(eu_pca_eval$eigenvalues))*100,digits=3),"%)")) +
  xlab(paste0("PC2 (",round((eu_pca_eval$eigenvalues[2]/sum(eu_pca_eval$eigenvalues))*100,digits=3),"%)")) +
  labs(shape = "Ancient Individual") +
  theme(legend.position="bottom") +
  guides(alpha=guide_legend(ncol=2))
```
\
```{r plot pca2 legend, echo=FALSE, message=FALSE, warning=FALSE, eval=nrow(pca_samples)>0}

## Prepare list of populations for text
eu_pops_for_text <- eu_pop_labels %>%
  group_by(region) %>%
  summarise(pops=paste(pop, collapse=", ")) %>%
  mutate(
    Region=case_when(
      region == "Caucasus" ~ "Caucasus",
      region == "NEast" ~ "Near East",
      region == "NEEur" ~ "Northeastern/Eastern Europe",
      region == "NEur" ~ "Central/North Europe",
      region == "SEur" ~ "South Europe"
    )
  ) %>%
  select(
    Region,
    Populations = pops
  )

eu_pops_for_text %>%
  kable(longtable = TRUE, format.args=list(big.mark = ',')) %>%
  kable_styling(latex_options = c("hold_position")) %>%
  column_spec(., 1, width="3.5cm") %>%
  column_spec(., 2, width="13cm") %>%
  row_spec(1, color="#00BF7D") %>%
  row_spec(2, color="#E76BF3") %>%
  row_spec(3, color="#00B0F6")
```
\else
No individuals pass the SNP coverage cutoff for this analysis (`r format(snp_cov_cutoff, digits = 0, big.mark = " ", decimal.mark = ",")` SNPs). A PCA for this data will be added here if/when additional sequencing data becomes available.
\fi
