---
output:
  pdf_document: default
  html_notebook:
    code_folding: hide
  html_document:
    df_print: paged
params: 
    set_subtitle: "Relic_samples"
    snp_coverage_file: "~/mount/eager_outputs/2020-05-03-relic_samples/genotyping/double_eigenstrat_coverage.txt"
    sex_det_file: "~/mount/eager_outputs/2020-05-03-relic_samples/sex_determination/SexDet.txt"
    stats_table: "~/mount/eager_outputs/2020-05-03-relic_samples/multiqc/multiqc_data/multiqc_general_stats.txt"
    poseidon_geno: "~/mount/poseidon_packages/2020-05-03-relic_samples/2020-05-03-relic_samples.geno.txt"
    poseidon_snp: "~/mount/poseidon_packages/2020-05-03-relic_samples/2020-05-03-relic_samples.snp.txt"
    poseidon_ind: "~/mount/poseidon_packages/2020-05-03-relic_samples/2020-05-03-relic_samples.ind.txt"
    poseidon_janno: "~/mount/forged_packages/microscope_pca/microscope_pca.janno"
    pmmr_results_fn: "~/mount/automated_analysis/2020-05-03-relic_samples/pmmr/2020-05-03-relic_samples.pmmr.txt"
    pca_bg_annotation: "~/Software/github/Schiffels-Popgen/MICROSCOPE-processing-pipeline/project_reports/assets/bg_annotation.txt"
    evec_fn: "~/mount/automated_analysis/microscope_pca/West_Eurasian_pca.evec"
    eval_fn: "~/mount/automated_analysis/microscope_pca/West_Eurasian_pca.eval"
    cred_file: "~/Software/github/Schiffels-Popgen/MICROSCOPE-processing-pipeline/.credentials"
    
title: "Extended preliminary data report"
subtitle: "`r params$set_subtitle`"
---

```{r load_libraries, message=FALSE, warning=FALSE, include=FALSE}
## Load required packages
library(dplyr)
library(tidyr)
library(ggplot2)
library(readr)
library(magrittr)
# library(knitr)
library(kableExtra)
library(forcats)
library(maps)
library(scales)
library(kableExtra)
library(ggrepel)
if(!require('remotes')) install.packages('remotes')
if(!require('sidora.core')) remotes::install_github("sidora-tools/sidora.core")
if(!require('poseidonR')) remotes::install_github('poseidon-framework/poseidonR')
library(sidora.core)
library(poseidonR)
require(RMariaDB)
# require(admixr)

## Cutoffs for genetic sex calls and expected ranges in sexdet plot.
#### Females
xx_xmin <- 0.7
xx_xmax <- 1.2
xx_ymin <- 0.0
xx_ymax <- 0.1
#### Males
xy_xmin <- 0.2
xy_xmax <- 0.6
xy_ymin <- 0.3
xy_ymax <- 0.6

sex_cutoffs <- tibble("Expected ranges"=c("XX", "XY"), 
                      xmin=c(xx_xmin, xy_xmin), 
                      xmax=c(xx_xmax, xy_xmax), 
                      ymin=c(xx_ymin, xy_ymin), 
                      ymax=c(xx_ymax, xy_ymax))

## Function to infer genetic sex from x/y rates
infer_sex <- function(sex_det, sex_cutoffs){
  ## Convert to a matrix with rownames so code is more legible and to avoid needing to pull() values from the tibbles.
  limits <- sex_cutoffs %>% tibble::column_to_rownames(var="Expected ranges") %>% as.matrix()
  sex_det <- sex_det %>%
    mutate("Genetic Sex" = ifelse(
      `x-rate` > limits["XX","xmin"] & 
        `x-rate` < limits["XX","xmax"] & 
        `y-rate` > limits["XX","ymin"] & 
        `y-rate` < limits["XX","ymax"], "F",
      ifelse(`x-rate` > limits["XY","xmin"] & 
               `x-rate` < limits["XY","xmax"] & 
               `y-rate` > limits["XY","ymin"] & 
               `y-rate` < limits["XY","ymax"], "M",
             "Unknown"))) %>%
    select(`Sample`, `Genetic Sex`)
  return(sex_det)
}
```

# Descriptive statistics of sequencing data

```{r load_general_stats, echo=FALSE, message=FALSE}
library_info <- read_tsv(params$stats_table) %>%
  filter(!grepl("_S0_",Sample)) %>% 
  filter(grepl("TF|SG",Sample)) %>% 
  select(Sample,
    "Sequenced Reads"=`Samtools Flagstat (pre-samtools filter)_mqc-generalstats-samtools_flagstat_pre_samtools_filter-flagstat_total`,
    "Mapped Reads" = `Samtools Flagstat (pre-samtools filter)_mqc-generalstats-samtools_flagstat_pre_samtools_filter-mapped_passed`,
    "Endogenous DNA (%)"= `endorSpy_mqc-generalstats-endorspy-endogenous_dna`,
    "% of Sequence Duplication" = `Picard_mqc-generalstats-picard-PERCENT_DUPLICATION`,
    "Damage 1st bp 3'" = `DamageProfiler_mqc-generalstats-damageprofiler-3_Prime1`
         ) %>%
  mutate(`% of Sequence Duplication`=`% of Sequence Duplication`*100) %>%
  arrange(Sample)
library_info %>% kable(format.args=list(big.mark = ',', digits=3))  %>% kable_styling(full_width=T)
```
\newpage

# Sex determination and SNP coverage

The table below includes the sample name, archaeological ID, genetic sex and SNP coverage of each individual. Genetic Sex is abbreviated as follows:

* M : Male
* F : Female
* Unknown (usually due to low coverage.)

Individuals that have too few covered SNPs (< 10,000) for population genetic analysis are shown in red.

```{r gather snp coverage data, echo=FALSE, message=FALSE}
## Load and parse coverage stats
snp_cov<-read_tsv(params$snp_coverage_file, col_names=c("#Sample", "#SNPs_Covered", "#SNPs_Total"), col_types='cii', skip=1) %>% 
  rename("Sample"="#Sample","# SNPs Covered" = "#SNPs_Covered", "Out of a maximum of" = "#SNPs_Total") %>%
  arrange(Sample) ## Sort by individual ID so bad_cov gets the correct indices for highlighting in the table.


bad_cov <- which(snp_cov$`# SNPs Covered` < 10000)
good_cov_inds <- snp_cov$Sample[which(snp_cov$`# SNPs Covered` >= 10000)]

```

```{r gather sexdet data, message=FALSE, include=FALSE}
## Load and parse sex determination stats.
sex_det <- read_tsv(params$sex_det_file, col_names=c("#Sample", "#SnpsAut", "#SNPsX", "#SnpsY", "NrAut", "NrX", "NrY", "x-rate", "y-rate"), col_types='ciiiiiidd', skip=1) %>%
  separate(`#Sample`, into=c("Sample"), sep="[._]", extra='drop') %>%
  select(`Sample`, `x-rate`, `y-rate`) %>%
  infer_sex(sex_cutoffs=sex_cutoffs)
```

```{r arch_IDs from pandora, message=FALSE, include=FALSE}
## Load ArchIDs from pandora using sidora
con <- get_pandora_connection(params$cred_file)
pandora_data <- get_df("TAB_Individual", con) %>% 
  filter(`individual.Full_Individual_Id` %in% sex_det$Sample) %>%
  select(`individual.Full_Individual_Id`, `individual.Archaeological_ID`) %>%
  rename("Sample" = `individual.Full_Individual_Id`, "Archaeological ID" = `individual.Archaeological_ID`)
RMariaDB::dbDisconnect(con)
```

```{r print coverage_sexdet_arch_id table, echo=FALSE, message=FALSE}
sex_det_pandora <- full_join(sex_det, pandora_data, by="Sample")
snp_cov_sex_pandora <- full_join(snp_cov, sex_det_pandora, by="Sample") %>% select(`Sample`, `Archaeological ID`, `Genetic Sex`, `# SNPs Covered`, `Out of a maximum of`) %>%
  arrange(Sample)

snp_cov_sex_pandora %>% kable(format.args=list(big.mark = ',')) %>% kable_styling(latex_options = "hold_position") %>%
  row_spec(bad_cov, bold = F, color = "red")
```
**These results are liable to change with any addition of further sequencing data.**
\newpage

The following figure shows the relative coverage of each individual on the X and Y chromosomes, and the associated error bars for each measure, as well as the expected position for males and females on such a plot. The expected ranges displayed here were used as the cutoffs for the inference of genetic sex shown in the table on page 2.

```{r sexdet plot, echo=FALSE, fig.height=7, fig.width=6, message=FALSE, warning=FALSE}
sex_det <- read_tsv(params$sex_det_file, col_names=c("#Sample", "#SnpsAut", "#SNPsX", "#SnpsY", "NrAut", "NrX", "NrY", "x-rate", "y-rate", "Err(x-rate)", "Err(y-rate)"), col_types='ciiiiiidddd', skip=1) %>%
  separate(`#Sample`, into=c("Sample"), sep="[._]", extra='drop') %>%
  mutate(xmin=`x-rate`-`Err(x-rate)`,
         xmax=`x-rate`+`Err(x-rate)`,
         ymin=`y-rate`-`Err(y-rate)`,
         ymax=`y-rate`+`Err(y-rate)`,
         `total_snps`=NrAut+NrY+NrX
  ) %>%
  select(Sample, x=`x-rate`, xmin, xmax, y=`y-rate`, ymin, ymax, "Total Snps"=total_snps)

ggplot(sex_det) +
  geom_point(aes(x=x, y=y, alpha=`Total Snps`), size=2.5) +
  geom_rect(data=sex_cutoffs, aes(xmin=xmin, xmax=xmax, ymin=ymin,ymax=ymax, fill=`Expected ranges`), alpha=0.4) +
  geom_errorbar(aes(x=x, ymin=ymin, ymax=ymax, alpha=`Total Snps`), width=0) +
  geom_errorbarh(aes(y=y, xmin=xmin, xmax=xmax, alpha=`Total Snps`), height=0) +
  geom_text_repel(aes(label=Sample, x=x, y=y), max.overlaps = 50, size=2) +
  scale_alpha(range=c(0.3,1), labels=function(x) {format(x, big.mark = ",", decimal.mark = ".", scientific = F)}) +
  theme_classic() +
  theme(legend.position = 'bottom', 
        legend.direction="horizontal", 
        legend.margin=margin(r=20,l=20), 
        legend.box='vertical'
        ) +
  xlab("Relative coverage on the X chromosome") +
  ylab("Relative coverage on the Y chromosome") +
  ggtitle("Genetic sex determination plot") +
  coord_cartesian(xlim=c(0,1.2), ylim=c(0,0.8)) + 
  expand_limits(x=c(0,1.2),y=c(0,0.8))
```

\newpage

# Pairwise mismatch rate between individuals
The pairwise genotype mismatch rate (pMMR) between different individuals, can be indicative of genetic kinship. Unrelated individuals will have a higher pMMR, with multiple unrelated pairs forming a plateau on the upper-right end of the figure. The pMMR of genetically related individuals is lowered proportionally to the closeness of their relationship. Some rough expected values for pMMR based on an underlying relationship are:

* Identical Twins/ Same individual: 50% of plateau value
* First Degree: 75% of plateau value
* Second Degree: 87.5% of plateau value

**It is often difficult to tell apart second from third degree relationships.** To do that, we often employ additional analytical methods once a suspected genetic relationship is identified with pMMR.

The colour of the points indicates the relative confidence around the pMMR. Red points are unreliable, gray points are somewhat reliable, and black points are reliable estimates of the pMMR. The numbers above each point correspond to the number of SNPs used in the calculation of the pMMR between each pair of individuals.

```{r import and plot pMMR, echo=FALSE, fig.height=5, fig.width=8.25, message=FALSE}
red_below_x_snps=100
black_above_x_snps=200

pmmr_data<-read_tsv(params$pmmr_results_fn) %>%
  mutate(Label=paste(Ind1,Ind2, sep="-"), colour=if_else(nSNPs < red_below_x_snps, paste0("<",red_below_x_snps," SNPs"), if_else(nSNPs < black_above_x_snps, paste0("<",black_above_x_snps," SNPs"),paste0(black_above_x_snps,"+ SNPs")))) %>%
  ## Remove rows with p Mismatch == NA which are invdividuals with 0 overlapping SNPs. Also remove rows with pMMR=0 or 1, which are caused by comparisons of individuals with low overlapping SNPs (usually less that 10). They can swamp other results in some datasets
  filter(!is.na(pMismatch), pMismatch > 0, pMismatch < 1)

## Dynamic figure size based on number of points.
fig_width=1+(0.2*nrow(pmmr_data))
fig_height=5/8.25*fig_width
```

```{r plot pMMR, echo=FALSE, fig.height=fig_height, fig.width=fig_width, message=FALSE}
if (nrow(pmmr_data) > 1) {
ggplot(pmmr_data, aes(x=fct_reorder(Label, pMismatch), y=pMismatch, col=factor(colour, levels=c(paste0("<",red_below_x_snps," SNPs"),paste0("<",black_above_x_snps," SNPs"),paste0(black_above_x_snps,"+ SNPs"))))) +
  scale_color_manual(values=stats::setNames(c("red","grey","black"), c(paste0("<",red_below_x_snps," SNPs"),paste0("<",black_above_x_snps," SNPs"), paste0(black_above_x_snps,"+ SNPs")))) +
  labs(color = "number of\noverlapping\nSNPs") +
  geom_point(size=2) + 
  geom_text(aes(label=nSNPs, y=pMismatch+0.05), size=1.5, angle=90) +
  theme_classic(base_size=13) + 
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=0.5), legend.title.align = 0.5) +
  expand_limits(x = 0, y = 0) +
  xlab('') +
  ylab("pairwise genotype mismatch rate")
} else {
  print("Pairwise mismatch rate cannot be calculated for a single individual.")
}
```
\newpage

# Principal component analysis

The following plot shows how ancient individuals  with sufficient genomic coverage (in black) relate to present-day West Eurasian populations. The distance between individuals roughly correlates with genetic distances. The percent of genetic variation explained by each axis is shown in parentheses.

The colour used in the plotting of present-day individuals indicates rough geographic regions, as follows:

* Caucasus: \textcolor[HTML]{F8766D}{red}
* Near East: \textcolor[HTML]{A3A500}{olive}
* South Europe: \textcolor[HTML]{00B0F6}{light blue}
* Central/North Europe: \textcolor[HTML]{E76BF3}{magenta}
* Northeaster/Eastern Europe: \textcolor[HTML]{00BF7D}{green}

```{r plot and plot PCA, echo=FALSE, fig.height=7, fig.width=7, message=FALSE, warning=FALSE}
pop_labels <- read_tsv(params$pca_bg_annotation, col_names = c("pop","label","region","colour"))
pca_evec <- readr::read_fwf(params$evec_fn, 
                          fwf_widths(
                            c(21,11,rep(12,times=9),17),
                              col_names=c("ind","pc1","pc2","pc3","pc4","pc5","pc6","pc7","pc8","pc9","pc10","pop")
                            ), 
                          skip=1, 
                          trim_ws=T,
                          col_select=c("ind","pc1","pc2","pop"), col_types='cddc'
                          )

bg_plot_data <- inner_join(pca_evec, pop_labels, by="pop")
pca_samples <- pca_evec %>% filter(ind %in% good_cov_inds)

pca_eval <- read_tsv(params$eval_fn, col_names = "eigenvalues")


## Infer PC orientation and flip if necessary
Chuvash_ind <- bg_plot_data %>% filter(ind == "Ttr-481")
BedouinB_ind <- bg_plot_data %>% filter(ind == "HGDP00636")
if (Chuvash_ind$pc1 > BedouinB_ind$pc1) {
  bg_plot_data <- bg_plot_data %>% mutate(pc1 = -pc1)
  pca_samples <- pca_samples %>% mutate(pc1 = -pc1)
}
if (Chuvash_ind$pc2 < BedouinB_ind$pc2) {
  bg_plot_data <- bg_plot_data %>% mutate(pc2 = -pc2)
  pca_samples <- pca_samples %>% mutate(pc2 = -pc2)
}

ggplot()+
  coord_equal() +
  geom_text(data=bg_plot_data, aes(x=pc1, y=pc2, label=label, col=factor(colour)), alpha=0.4, size=2, show.legend=F) +
  theme_minimal() +
  geom_point(data=pca_samples, aes(x=pc1, y=pc2)) +
  geom_text_repel(data=pca_samples, aes(x=pc1, y=pc2, label=ind), max.overlaps = Inf, size=2.5) +
  # geom_point(data=pca_samples, aes(x=pc1, y=pc2, shape=ind)) +
  xlab(paste0("PC1 (",round((pca_eval$eigenvalues[1]/sum(pca_eval$eigenvalues))*100,digits=3),"%)")) +
  ylab(paste0("PC2 (",round((pca_eval$eigenvalues[2]/sum(pca_eval$eigenvalues))*100,digits=3),"%)")) +
  labs(shape = "Ancient Individual")
  
```
